### PR TITLE
drivers: i2c: Introduce I2C timeout for SAM0

### PR DESCRIPTION
**Pull request by kowalewskijan**
Adds configurable timeout for I2C transactions and provides bus recovery functionality.

Tested on SAME53J20A

### CODE REVIEW

File: drivers/i2c/i2c_sam0.c [hunk 1]
```diff
@@ -60,6 +85,50 @@ struct i2c_sam0_dev_data {
...
+	gpio_pin_configure(sda_dev, sda_pin, GPIO_OUTPUT | GPIO_OUTPUT_INIT_HIGH);
+	gpio_pin_set(sda_dev, sda_pin, 0);
+	k_usleep(4);
+	gpio_pin_set(sda_dev, sda_pin, 1);
+	k_usleep(4);
+
+	/* Recover pinmux */
+	pinctrl_configure_pins(&soc_pin_sda, 1, PINCTRL_REG_NONE);
+	pinctrl_configure_pins(&soc_pin_scl, 1, PINCTRL_REG_NONE);
+
```
Comment (nandojve) [inline drivers/i2c/i2c_sam0.c:128]: Why do you need this? Why not use below?

```
	retval = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
	if (retval < 0) {
		return retval;
	}
```

File: drivers/i2c/i2c_sam0.c [hunk 2]
```diff
@@ -60,6 +85,50 @@ struct i2c_sam0_dev_data {
...
+	/* Save original pin config values */
+	soc_pin_sda = cfg->pcfg->states[PINCTRL_STATE_DEFAULT].pins[SERCOM_I2CM_SDA_PIN_IDX];
+	soc_pin_scl = cfg->pcfg->states[PINCTRL_STATE_DEFAULT].pins[SERCOM_I2CM_SCL_PIN_IDX];
+
+	uint8_t sda_port_idx = SAM_PINMUX_PORT_GET(soc_pin_sda);
+	uint8_t scl_port_idx = SAM_PINMUX_PORT_GET(soc_pin_scl);
+	uint32_t sda_pin = SAM_PINMUX_PIN_GET(soc_pin_sda);
+	uint32_t scl_pin = SAM_PINMUX_PIN_GET(soc_pin_scl);
+	const struct device *sda_dev = sam_port_devs[sda_port_idx];
+	const struct device *scl_dev = sam_port_devs[scl_port_idx];
```
Comment (nandojve) [inline drivers/i2c/i2c_sam0.c:102]: Could you pointing out an example where you found different ports for I2C pins at same?

**Review by nandojve — CHANGES_REQUESTED** [review]
Hi @kowalewskijan ,

Interesting, a few points:

1- Could you explain on the commit messages what you are solving and why? It will be nice that you add information about a few use cases too. In this case, are you trying to solve some of the problems related on **36.6.2.4.2 Transmitting Address Packets**?

2- I can imagine that it is not possible to have a HW test to avoid regressions for this, right?

3- Zephyr is moving to have all dependencies defined by each driver. This means that you need to add GPIO and PINCTRL dependencies at Kconfig.

**Comment by kowalewskijan** [comment]
@nandojve, what I mainly want to add is the timeout for I2C in sam0 driver. Right now, once the I2C is blocked due to malfunctioning I2C slave device or any other hardware issue that may cause the I2C to be not functional, the driver will hang the whole system during the device init phase. I believe introducing timeouts in such case will greatly improve usability for these SoCs. Regarding recovering the I2C it's just a typical way to recover it, but thinking about it, maybe it should handle the I2C errors more holistic based on STATUS register. If you want, we can split this into 1st having the timeout which is definitely beneficial for overall stability of the system for these SoCs and later add recovery features based on STATUS register.

**Comment by nandojve** [comment]
Hi @kowalewskijan ,

> @nandojve, what I mainly want to add is the timeout for I2C in sam0 driver. Right now, once the I2C is blocked due to malfunctioning I2C slave device or any other hardware issue that may cause the I2C to be not functional, the driver will hang the whole system during the device init phase. I believe introducing timeouts in such case will greatly improve usability for these SoCs. Regarding recovering the I2C it's just a typical way to recover it, but thinking about it, maybe it should handle the I2C errors more holistic based on STATUS register. If you want, we can split this into 1st having the timeout which is definitely beneficial for overall stability of the system for these SoCs and later add recovery features based on STATUS register.

I like the idea about split into 2 parts. I'm fine about timeouts unless @teburd have a different opinion. What I really appreciate is to receive a solution that uses the STATUS registers to perform the bus recover.

**Comment by teburd** [comment]
> Hi @kowalewskijan ,
> 
> > @nandojve, what I mainly want to add is the timeout for I2C in sam0 driver. Right now, once the I2C is blocked due to malfunctioning I2C slave device or any other hardware issue that may cause the I2C to be not functional, the driver will hang the whole system during the device init phase. I believe introducing timeouts in such case will greatly improve usability for these SoCs. Regarding recovering the I2C it's just a typical way to recover it, but thinking about it, maybe it should handle the I2C errors more holistic based on STATUS register. If you want, we can split this into 1st having the timeout which is definitely beneficial for overall stability of the system for these SoCs and later add recovery features based on STATUS register.
> 
> I like the idea about split into 2 parts. I'm fine about timeouts unless @teburd have a different opinion. What I really appreciate is to receive a solution that uses the STATUS registers to perform the bus recover.

I don't have a strong opinion on this myself. I trust your judgment here @nandojve

File: drivers/i2c/i2c_sam0.c (hunk 1 above)
Comment (kowalewskijan) [reply drivers/i2c/i2c_sam0.c]: removed from this PR

**Comment by kowalewskijan** [comment]
@nandojve  - I extracted only I2C timeouts for this PR

**Review by nandojve — APPROVED** [review]

**Review by pdgendt — APPROVED** [review]

**Review by teburd — APPROVED** [review]

**Pull request merged**